lib/jido_action/schema.ex

```diff
@@ -0,0 +1,264 @@
+  def to_json_schema([]), do: %{}
+  def to_json_schema(schema) when is_list(schema), do: nimble_to_json_schema(schema)
+  def to_json_schema(schema), do: Zoi.to_json_schema(schema)
+
+  @doc """
+  Formats validation errors into Jido.Action.Error structs.
+
+  ## Parameters
+    * `error` - The error from validation (NimbleOptions.ValidationError, Zoi.Error, or list)
+    * `context` - Context string describing where the error occurred
+    * `module` - The module where the error occurred
+
+  ## Returns
+    * `Jido.Action.Error.InvalidInputError.t()` - Formatted error struct
+  """
+  @spec format_error(term(), String.t(), module()) ::
+          Jido.Action.Error.InvalidInputError.t()
+  def format_error(error, context, module) do
+    case error do
+      %NimbleOptions.ValidationError{} = nimble_error ->
+        nimble_error
+        |> Jido.Action.Error.format_nimble_validation_error(context, module)
+        |> Jido.Action.Error.validation_error()
+
+      %Zoi.Error{} = zoi_error ->
+        format_zoi_error(zoi_error, context, module)
+
+      errors when is_list(errors) ->
+        message = Zoi.prettify_errors(errors)
+
+        Jido.Action.Error.validation_error(message, %{
+          context: context,
+          module: module,
+          errors: format_zoi_error_list(errors)
+        })
+
+      _ ->
+        Jido.Action.Error.validation_error("Validation failed", %{
+          context: context,
+          module: module
+        })
+    end
+  end
+
+  @doc """
+  Validates a schema value for use in configuration.
+
+  Used during compilation to ensure schema configuration is valid.
+
+  ## Parameters
+    * `value` - The schema value to validate
+    * `_opts` - Options (unused, for Zoi refine compatibility)
+
+  ## Returns
+    * `:ok` - Schema is valid
+    * `{:error, message}` - Schema is invalid
+  """
+  @spec validate_config_schema(term(), keyword()) :: :ok | {:error, String.t()}
+  def validate_config_schema(value, _opts \\ [])
+
+  def validate_config_schema(value, _opts) when is_list(value), do: :ok
+
+  def validate_config_schema(value, _opts) do
+    if impl_for_zoi_type?(value) do
+      :ok
+    else
+      {:error, "must be NimbleOptions schema or Zoi schema"}
+    end
+  end
+
+  # Private Functions
+
+  defp impl_for_zoi_type?(value) do
+    is_struct(value) && Zoi.Type.impl_for(value) != nil
+  rescue
+    _ -> false
+  end
+
+  defp validate_nimble(schema, data) do
+    data_kw = if is_map(data), do: Enum.to_list(data), else: data
+
+    case NimbleOptions.validate(data_kw, schema) do
+      {:ok, validated_kw} -> {:ok, Map.new(validated_kw)}
+      {:error, error} -> {:error, error}
+    end
+  end
+
+  defp validate_zoi(schema, data) do
+    case Zoi.parse(schema, data) do
+      {:ok, validated} -> {:ok, validated}
+      {:error, errors} -> {:error, errors}
+    end
+  end
+
+  defp extract_zoi_keys(%{__struct__: Zoi.Types.Object, fields: fields}) when is_map(fields) do
```

Comment:
can do this in a single clause:

```elixir
  defp extract_zoi_keys(%{fields: fields}) do
    Enum.map(fields, fn {k, _v} -> k end)
  end
```

since map/lists have the enumerable protocol implemented